### PR TITLE
kata-deploy: Add Helm Chart

### PR DESCRIPTION
**Pull request by zvonkok**
For easier handling of kata-deploy we can leverage a Helm chart to get rid of all the base and overlays for the various components

### CODE REVIEW

**Comment by zvonkok** [comment]
Per default the appVersion would be set to VERSION one can override it by simply saying: 

```
helm install --set image.tag=latest  kata-deploy-0.1.0.tgz
``` 
A default 
```
helm install ./kata-deploy-0.1.0.tgz 
``` 
would give one the latest release.

**Comment by zvonkok** [comment]
The helm-chart can be eaisly hosted via github.io on kata-containers.

**Comment by zvonkok** [comment]
With this we can also automate the chart publishing for each release https://github.com/helm/chart-releaser

File: tools/packaging/kata-deploy/helm-chart/Makefile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+
+VERSION_FILE := ../../../../VERSION
+
+release:
+	sed -i 's/appVersion: .*/appVersion: $(shell cat $(VERSION_FILE))/g' kata-deploy/Chart.yaml
+all: package
+
+package: helm release
```
Comment (beraldoleal) [inline tools/packaging/kata-deploy/helm-chart/Makefile:8]: I was expecting a "helm install" somewhere in our CI webhooks/tests, that would consume this package.

So do you mind expanding a little bit on the plan? What is next? Who would be calling this target and who would be consuming the `kata-deploy-0.1.0.tgz` helm package ?

Your commit message says "get all the base and overlays", should we try to replace it now within this PR?

**Review by beraldoleal — COMMENTED** [review]
Hi @zvonkok , thanks a lot for this. IMO a good improvement. Just left a few questions for you.

File: tools/packaging/kata-deploy/helm-chart/Makefile (hunk 1 above)
Comment (zvonkok) [reply tools/packaging/kata-deploy/helm-chart/Makefile:8]: Yes that is the idea to use the helm chart for our CI and for our releases for users to have a pinned version to install from. I did not want to overload this PR just kicking of the discussion and see where and how to extend this.

**Comment by zvonkok** [comment]
@beraldoleal Take a look at the second commit how we can  use Helm for rendering the correct yamls without changing the yamls in place and hence making the repository dirty. 

FYI @fidencio @ryansavino

File: tests/integration/kubernetes/gha-run.sh [hunk 2]
```diff
@@ -157,83 +158,64 @@ function deploy_kata() {
...
 	echo "::endgroup::"
 
-	kubectl_retry apply -f "${tools_dir}/packaging/kata-deploy/kata-rbac/base/kata-rbac.yaml"
-	case "${KUBERNETES}" in
-		k0s) kubectl_retry apply -k "${tools_dir}/packaging/kata-deploy/kata-deploy/overlays/k0s" ;;
-		k3s) kubectl_retry apply -k "${tools_dir}/packaging/kata-deploy/kata-deploy/overlays/k3s" ;;
-		rke2) kubectl_retry apply -k "${tools_dir}/packaging/kata-deploy/kata-deploy/overlays/rke2" ;;
-		*) kubectl_retry apply -f "${tools_dir}/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml"
-	esac
+	helm install kata-deploy ./"${helm_chart_dir}" --set-file values.yaml="${values_yaml}" --namespace kube-system
```
Comment (zvonkok) [inline tests/integration/kubernetes/gha-run.sh:218]: @beraldoleal Here is "your" install :) using this method we also apply any helm chart changes to the CI

File: tools/packaging/kata-deploy/helm-chart/Makefile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@[ -f $(PROJECT_DIR)/bin/$(1) ] || command -v $(1) >/dev/null 2>&1 || { \
+set -e ;\
+echo "Downloading $(2)" ;\
+GOBIN=$(PROJECT_DIR)/bin go install $(2) ;\
+rm -rf $$TMP_DIR ;\
+}
+endef
+
+HELM = $(shell pwd)/bin/helm
+helm: ## Download helm locally if necessary.
```
[the remark this answers is not in the document]
Comment (zvonkok) [reply tools/packaging/kata-deploy/helm-chart/Makefile:28]: I will remove this. Consider this a NOOP.

**Comment by zvonkok** [comment]
kata-deploy.yaml and kata-cleanup.yaml are the same manifests and scripts just with different arguments next commit will clean this up

**Comment by zvonkok** [comment]
`cleanup_kata_deploy` is now really simple, just uninstall the helm release if found. All the needed data is encapsulated in the deployed release.

**Comment by zvonkok** [comment]
With this we should be able to remove "ALL" base/overlay kustomzie manifests.

**Review by ryansavino — COMMENTED** [review]
Nice. Made a few comments. I think this would help fix several issues that we brought up in the ci meeting.

Specifically:
https://github.com/kata-containers/kata-containers/issues/9793

This PR eliminates some of the longer sleep times that were mentioned in the following issues.
https://github.com/kata-containers/kata-containers/issues/9795
https://github.com/kata-containers/kata-containers/issues/9796

File: tests/integration/kubernetes/gha-run.sh [hunk 4]
```diff
@@ -157,83 +158,64 @@ function deploy_kata() {
...
+	  agentNoProxy: "${AGENT_NO_PROXY}"
+	  pullTypeMapping: "${PULL_TYPE_MAPPING}"
+	  hostOS: "${HOST_OS}"
+	EOF
+
 	echo "::group::Final kata-deploy.yaml that is used in the test"
-	cat "${tools_dir}/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml"
-	grep "${DOCKER_REGISTRY}/${DOCKER_REPO}:${DOCKER_TAG}" "${tools_dir}/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml" || die "Failed to setup the tests image"
+	helm template ./"${helm_chart_dir}" --set-file values.yaml="${values_yaml}" --namespace kube-system
+	sed -n "/reference: ${DOCKER_REGISTRY}\/${DOCKER_REPO}/{N;/\n[[:space:]]*tag: ${DOCKER_TAG}/p;}" "${values_yaml}"
```
[the remark this answers is not in the document]
Comment (zvonkok) [reply tests/integration/kubernetes/gha-run.sh:215]: sed is used here to match a multi-line pattern not to replace.

File: tests/common.bash [hunk 5]
```diff
@@ -511,6 +511,24 @@ function ensure_yq() {
 	hash -d yq 2> /dev/null || true # yq is preinstalled on GHA Ubuntu 22.04 runners so we clear Bash's PATH cache.
 }
 
+function ensure_helm() {
+	# Check if helm is available in the system's PATH
+	if ! command -v helm &> /dev/null; then
+		echo "Helm is not installed. Installing Helm..."
+		# Use curl method to install Helm
+		curl -fsSL https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 | bash
```
[the remark this answers is not in the document]
Comment (zvonkok) [reply tests/common.bash:519]: That is a good point. Let's pin the helm version.

File: tests/integration/kubernetes/gha-run.sh [hunk 6]
```diff
@@ -157,83 +158,64 @@ function deploy_kata() {
...
 	fi
 
 	if [ "${KATA_HOST_OS}" = "cbl-mariner" ]; then
-		yq -i \
-		  '.spec.template.spec.containers[0].env[6].value = "initrd kernel default_vcpus"' \
-		  "${tools_dir}/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml"
-		yq -i \
-		  ".spec.template.spec.containers[0].env += [{\"name\": \"HOST_OS\", \"value\": \"${KATA_HOST_OS}\"}]" \
-		  "${tools_dir}/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml"
+		ALLOWED_HYPERVISOR_ANNOTATIONS="image initrd kernel default_vcpus"
```
[the remark this answers is not in the document]
Comment (zvonkok) [reply tests/integration/kubernetes/gha-run.sh:173]: Copy and waste, good catch!

**Comment by zvonkok** [comment]
I've build a simple GHA job to upload a Helm chart for each PR automatically to a specific repository. In my case its `zvonkok/helm-charts` those are hosted via github.io. 

Replace here `zvonkok/helm-charts` with `kata-containers/helm-charts` moving forward just using `zvonkok/helm-charts` for illustration purposes. 

```
helm repo add kata-deploy https://zvonkok.github.io/helm-charts
helm repo update
helm search repo --devel -o json 
[{"name":"kata-deploy/kata-deploy","version":"3.6.0-dev+24-4d45fd9818726d7d3a37cfd4ad1281bea29b67c2","app_version":"3.6.0-dev+24-4d45fd9818726d7d3a37cfd4ad1281bea29b67c2","description":"A Helm chart for deploying Kata Containers"}]
```
The `input.tag` (pr-githash) is defined by the GHA and I am using the VERSION + the tag to create a new sermer version for Helm to consume.

**Comment by zvonkok** [comment]
Additionally `values.yaml` would be updated for the chart to point to the correct payload 
```
imagePullPolicy: Always
imagePullSecrets: []
image:
  reference: ghcr.io/zvonkok/kata-deploy-ci/kata-deploy
  tag: 24-b9d7f9333087e5ee789af8f37dd26df3b3e308e0
# k8s-dist can be k8s, k3s, rke2, k0s
k8sDistribution: "k8s"
env:
  debug: "false"
  shims: "clh cloud-hypervisor dragonball fc qemu qemu-coco-dev qemu-runtime-rs qemu-sev qemu-snp qemu-tdx stratovirt qemu-nvidia-gpu qemu-nvidia-gpu-snp qemu-nvidia-gpu-tdx"
  defaultShim: "qemu"
  createRuntimeClasses: "false"
  createDefaultRuntimeClass: "false"
  allowedHypervisorAnnotations: ""
  snapshotterHandlerMapping: ""
  agentHttpProxy: ""
  agentNoProxy: ""
  pullTypeMapping: ""
  hostOS: ""
```

**Comment by zvonkok** [comment]
For each PR we woudl have a Helm chart `kata-deploy-VERSION-dev+{{ input.tag }},tar.gz` and for each release we would have a `kata-deploy-VERSION.tar.gz` with updated values.yaml also pushed as an artifact fo the release payload. 

Users can then do a `helm repo update` and without the `--devel` flag they will not see any  `kata-deploy-VERSION-dev+{{ input.tag }},tar.gz`  charts, only the release charts.

**Comment by ryansavino** [comment]
I was looking through the failed test runs for the amd node jobs. Let me know if you want to spend some time troubleshooting the failures together. Looks like it wasn't able to pull the kata-deploy image.

**Comment by zvonkok** [comment]
@ryansavino Found the error. Thanks for the offer :)

File: tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+imagePullPolicy: Always
+imagePullSecrets: []
+image:
+  reference: quay.io/kata-containers/kata-deploy
+  tag: ""
```
Comment (zvonkok) [inline tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml:5]: If tag is empty we will per default use chart.version which translates  to VERSION

**Review by mkulke — COMMENTED** [review]
does it make sense to add a `values.schema.json` to the chart? In my experience this takes the edge off when working with helm charts.

**Comment by zvonkok** [comment]
@mkulke Good point, adding this to the list of follow up items: 
https://github.com/kata-containers/kata-containers/issues/9924

**Comment by BbolroC** [comment]
Please rebase this PR onto main when you want to re-trigger the whole set of checks (by pushing something and etc.) as #9923 resolves the issue for the zvsi tests. Thanks.

**Comment by fidencio** [comment]
@zvonkok, it's taking me some to get to this, but I'd like to ensure this works with TDX first.
I will be force-pushing to your branch.

**Comment by fidencio** [comment]
I forced-pushed here to rebase.

**Review by fidencio — APPROVED** [review]
TDX CI is failing but other TEEs are passing.

lgtm, thanks @zvonkok!

**Comment by fidencio** [comment]
Found the issue, agent.https_proxy is not being properly set!

**Comment by fidencio** [comment]
```
git diff
diff --git a/tools/packaging/kata-deploy/helm-chart/kata-deploy/templates/kata-deploy.yaml b/tools/packaging/kata-deploy/helm-chart/kata-deploy/templates/kata-deploy.yaml
index 714e172e44..0d3565da38 100644
--- a/tools/packaging/kata-deploy/helm-chart/kata-deploy/templates/kata-deploy.yaml
+++ b/tools/packaging/kata-deploy/helm-chart/kata-deploy/templates/kata-deploy.yaml
@@ -47,7 +47,7 @@ spec:
         - name: SNAPSHOTTER_HANDLER_MAPPING
           value: {{ .Values.env.snapshotterHandlerMapping | quote }}
         - name: AGENT_HTTPS_PROXY
-          value: {{ .Values.env.agentHttpProxy | quote }}
+          value: {{ .Values.env.agentHttpsProxy | quote }}
         - name: AGENT_NO_PROXY
           value: {{ .Values.env.agentNoProxy | quote }}
         - name: PULL_TYPE_MAPPING
diff --git a/tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml b/tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml
index 004137a147..b1f195d1f1 100644
--- a/tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml
+++ b/tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml
@@ -13,7 +13,7 @@ env:
   createDefaultRuntimeClass: "false"
   allowedHypervisorAnnotations: ""
   snapshotterHandlerMapping: ""
-  agentHttpProxy: ""
+  agentHttpsProxy: ""
   agentNoProxy: ""
   pullTypeMapping: ""
   hostOS: ""
```

This will solve the issue.

**Comment by fidencio** [comment]
Force pushed, probably fixing the issue.

**Review by fidencio — APPROVED** [review]
lgtm, now with the TDX tests also passing.

Thanks, @zvonkok!

**Comment by ryansavino** [comment]
For snp, the `k8s-policy-hard-coded.bats` test keeps failing. Looks like the pods aren't starting. I think this is probably unrelated, but I'd like to troubleshoot and diagnose a bit further. The nightly CI seems to be passing fine.

Do you think a rebase may help here?

**Comment by fidencio** [comment]
> For snp, the `k8s-policy-hard-coded.bats` test keeps failing. Looks like the pods aren't starting. I think this is probably unrelated, but I'd like to troubleshoot and diagnose a bit further. The nightly CI seems to be passing fine.

What happened here was that the PR was rebased before the commit adding the test was added, and at that time the SNP CI was broken, with failures happening even before starting the tests.  You guys fixed the issue there, but meanwhile the test you mentioned was merged, and the auto-rebase would pick that up as part of the tests to run, and then we ended up with that test failing.

I've rebased now, and this should give us a green CI everywhere.

**Comment by ryansavino** [comment]
> > For snp, the `k8s-policy-hard-coded.bats` test keeps failing. Looks like the pods aren't starting. I think this is probably unrelated, but I'd like to troubleshoot and diagnose a bit further. The nightly CI seems to be passing fine.
> 
> What happened here was that the PR was rebased before the commit adding the test was added, and at that time the SNP CI was broken, with failures happening even before starting the tests. You guys fixed the issue there, but meanwhile the test you mentioned was merged, and the auto-rebase would pick that up as part of the tests to run, and then we ended up with that test failing.
> 
> I've rebased now, and this should give us a green CI everywhere.

Great. Thanks for explaining that. I was a bit confused. Approving.

**Review by ryansavino — APPROVED** [review]

**Pull request merged**